### PR TITLE
Streams: merging, mapping, whatnots

### PR DESCRIPTION
**Pull request by amosr**

### CODE REVIEW

**Comment by tranma** [comment]
I tried replacing the lazy Maybes in `merge` with strict Maybes.

- lazy: time 541.60 secs, alloc 1,182,763,895,920 bytes
- strict: time 134.71 secs, alloc  344,405,731,256 bytes

**Comment by amosr** [comment]
Ah, SpecConstr isn't getting rid of the Maybes: that explains it. Will have another try, not sure why it's not happening

File: x-vector/src/X/Data/Vector/Stream/Merge.hs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+------------|-------------|---------|---------------|-------------|---------------|
+            |    Just     | merge   | read-R        | fill-R      | read-R        |
+ Nothing    | ------------|---------|---------------|-------------|---------------|
+            |   Nothing   | merge   | fill-L        | fill-R      | done          |
+------------|-------------|---------|---------------|-------------|---------------|
+
+-}
+merge :: Monad m => (a -> a -> MergePullFrom a) -> VS.Stream m a -> VS.Stream m a -> VS.Stream m a
+merge f (VS.Stream l'step l'state) (VS.Stream r'step r'state)
+ = VS.Stream go'step (Just l'state, Just r'state, Nothing, Nothing)
```
Comment (tranma) [inline x-vector/src/X/Data/Vector/Stream/Merge.hs]: I think this might need to definitely be inlined for SpecConstr to fire

**Comment by amosr** [comment]
is this alright? @tranma

**Comment by tranma** [comment]
Yep, this looks fine to me 🎂 
But it would be nice to sound an alarm if the benchmarks change, since we want this exact inlining behaviour. Not sure how and if other projects are doing this?

**Comment by amosr** [comment]
I agree, I'm not sure how to do that though

**Comment by tranma** [comment]
strings?

**Comment by thumphries** [comment]
> Not sure how and if other projects are doing this?

I don't think anything does this right now, most benchmark suites are disabled (numbers are always severely inflated on CI)

**Comment by olorin** [comment]
> I don't think anything does this right now, most benchmark suites are disabled (numbers are always severely inflated on CI)

Yeah, sounds like a recipe for pain if implemented before we have dedicated benchmark boxes/some other way to run isolated builds.

**Comment by amosr** [comment]
yeah, let's just put the onus on whoever modifies x-vector next

**Pull request merged**